src/client.ts

```diff
@@ -343,23 +353,46 @@ class OptimizelyReactSDKClient implements ReactSDKClient {
   }
 
   public async fetchQualifiedSegments(options?: optimizely.OptimizelySegmentOption[]): Promise<boolean> {
-    if (!this.userContext || !this.isReady()) {
+    if (this.odpExplicitlyOff) {
+      return true;
+    }
+
+    if (!this.userContext) {
       return false;
     }
 
     return await this.userContext.fetchQualifiedSegments(options);
   }
 
   public async setUser(userInfo: UserInfo): Promise<void> {
-    this.setCurrentUserContext(userInfo);
-
     this.user = {
       id: userInfo.id || DefaultUser.id,
       attributes: userInfo.attributes || DefaultUser.attributes,
     };
 
-    if (this.getIsReadyPromiseFulfilled()) {
-      await this.fetchQualifiedSegments();
+    let fetchQualifedSegmentsSucceed = false;
+    if (userInfo.id === DefaultUser.id) {
+      // if user is anonymous...
+      // wait for the SDK client to be ready before
+      await this._client?.onReady();
+      // setting the user context
+      this.setCurrentUserContext(userInfo);
+
+      this.user.id = this.userContext?.getUserId() || DefaultUser.id;
```

Comment:
Can we add comment why we update id again here?

Reply:
Updating...